c++/src/ColumnWriter.cc

```diff
@@ -1837,7 +1837,7 @@ namespace orc {
         // TimestampVectorBatch already stores data in UTC
         int64_t millsUTC = secs[i] * 1000 + nanos[i] / 1000000;
         if (!isUTC) {
-          millsUTC = timezone.convertToUTC(millsUTC);
+          millsUTC = timezone.convertToUTC(secs[i]) * 1000 + nanos[i] / 1000000;
```

Comment:
Thank you, @noirello . Could you make a test case for this fix?

Reply:
Sure, added a UTC and a non UTC test case.

Reply:
Please, double check the timestamps in the test cases. Time zones always can be confusing.

Reply:
It's a little counter-intuitive because we only convert `secs[i]` only. Could you add some comment here about why we don't need to convert `nanos[i]`?

Reply:
cc @wgtmac , @stiga-huang too because this is a correctness issue.

Reply:
If we have set nanosecond in the test case _**WriterTest.writeTimestampWithTimezone**_, this issue may be fixed earlier.

Reply:
> Yes, it's a little counter-intuitive because we only convert `secs[i]` only. So, `Timezone.convertToUTC` doesn't care `nanos` part properly and we should not put it here?
> 
> https://github.com/apache/orc/blob/f4c7cc14b31876e3dd4934b4bfc40c05d4b8ebca/c%2B%2B/src/Timezone.cc#L604-L606
> 
> If then, can we fix it in `Timezone.convertToUTC` instead? Is there a side-effect?

I think the current fix is good enough. As time_t uses second internally, we'd better keep the contract of Timezone.convertToUTC. We may add some comment above the fix to help understanding.

Reply:
Got it. +1 for @wgtmac 's final decision.